corehq/apps/data_dictionary/util.py

```diff
@@ -0,0 +1,52 @@
+from corehq.apps.app_manager.dbaccessors import get_case_types_from_apps
+from corehq.apps.data_dictionary.models import CaseProperty, CaseType
+from corehq.apps.export.models.new import CaseExportDataSchema
+
+
+def generate_data_dictionary(domain):
+    current_properties = {}
+    if CaseType.objects.filter(domain=domain).count() != 0:
+        props = CaseProperty.objects.filter(case_type__domain=domain).prefetch_related('case_type')
+        for prop in props:
+            case_type = prop.case_type.name
+            if case_type not in current_properties:
+                current_properties[case_type] = {prop.name}
+            else:
+                current_properties[case_type].add(prop.name)
+
+    properties = _get_all_case_properties(domain)
+    new_case_properties = []
+    for case_type, properties in properties.items():
+        if not case_type:
+            continue
+
+        case_type_obj, _ = CaseType.objects.get_or_create(domain=domain, name=case_type)
```

Comment:
could keep a dict of case types fetched above and only do this call if we don't already have the case type (which is only if the case type doesn't exist or has no properties)